Attitude topics cleanup

mhkabir opened the pull request:
- Remove estimator low pass filtering of `vehicle_attitude` angular rates. (Needs discussion about whether the the rates are currently LPF'ed before being integrating to 250Hz)
- Remove `control_state` topic.
- Add minimal `sensors_corrected` topic for controllers to have access to in-run bias estimates and other consumers (e.g land detector, visual inertial odometry) to have access to *bias and thermal compensated* sensor measurements.

TODO : 
- ~Finish integration of sensor_corrected~
- ~Add back baro altitude in global position.~
- ~Make attitude_estimator_q publish `sensor_corrected`.~
- ~Fix Flight Review to use correct topics for groundspeed and airspeed.~
- ~`HIGHRES_IMU` should use the un-biased readings when available.~

mhkabir commented:
@priseborough Could you please review again? I split out the EKF's estimated rate bias from the attitude topic into `sensors_corrected` and addressed the concerns you had on the old PR.

For clarity, `sensor_corrected` includes the EKF's in-run bias estimates for the current sensor set and corrected sensor values for other consumers (visual inertial odometry, land detector, launch detection)

priseborough commented:
I'm working through these changes to get the big picture view on what has been done, but because of the number of modules that have been touched, this will need to be thoroughly tested. Do you have a plan on how you are going to achieve adequate test coverage?

Can we get this back to back tested against master on a fixed wing and multi-copter platform so we can review the effect on control loop noise and the landing detectors?

mhkabir commented:
@priseborough The majority of the changes are non-functional (cleanup and/or rearrangement). The functional and most critical bit would be the removal of the low pass filters in the estimators which was applied to the rate data : https://github.com/PX4/Firmware/pull/7523/commits/d3c7d47f2c8c4750abb91d92335f9722aa9960db

We'll need adequate testing with master side-by-side to see the effects and performance differences.

Also, regarding this comment on the PR thread : 

> That will depend on what has been done to the gyro data upstream. If It has been sampled at a high rate, and LPF before downsampling to 250Hz, than additional filtering at 250Hz is unnecessary and also does nothing for aliasing protection.

@dagar @bkueng @jgoppert I don't believe that we currently LPF the raw sensor data before down-sampling it?

mhkabir commented:
SITL Logs : 
- PR : http://logs.px4.io/plot_app?log=cf34d9fd-9608-444b-89ec-9fd5e18c6450
- Master : http://logs.px4.io/plot_app?log=2acd6378-43a3-4669-9bb3-fe94b3641ce9

jgoppert commented:
@mhkabir currently there is a 2nd order butterworth running in the imu driver at 1 kHz that low passes the imu data at 30 hz.

mhkabir commented:
@PX4TestFlights Please test on Fixedwing and Multicopter. We need direct comparisons against master, so you should generate 2 flights per vehicle (as similar as possible), on master and on this branch.

Avysuarez commented:
Flight running V3 (pixhawk mini); good flight, no issues.
http://review.px4.io/plot_app?log=20232e73-4e19-483b-850f-8c498886e9b8

dagar commented:
@Avysuarez is this the same vehicle flying master? http://review.px4.io/plot_app?log=07fc2a7d-209a-4893-a52f-25862460e5ee

If so it's nothing to get excited about, but it does lower cpu a bit. I would guess mostly from logger. Curious that your logging dropouts seem so regular and consistent between flights.

santiago3dr commented:
few FW flight on the pixhawk mini (V3) using a Phantom Flying Wing
running stable firmware:
http://logs.px4.io/plot_app?log=ed730136-511f-4a85-9b30-5f303a9815f3
Master:
http://logs.px4.io/plot_app?log=7d53d07c-3eda-4173-8443-9543ebf7ff13
http://logs.px4.io/plot_app?log=cd5b3c41-ffe8-4b47-a39f-18c1cfd9dab0
PR_7523:
http://logs.px4.io/plot_app?log=1df87b2b-e268-4521-b59d-f81b1fac6d3d
http://logs.px4.io/plot_app?log=23600d2b-694e-4aa5-8e6f-4cedb03e2b76

all in all, good flight with this PR; no notable difference with master

priseborough commented:
It is normal with PX4 to see loss of inertial data to the EKF when:

- Logging is started
- Parameters are saved

That is additional to log dropouts that occur if the logger RAM allocation is insufficient to buffer data when the sdcard periodically stops writing.

dagar commented:
@priseborough param save should be less of an issue now - https://github.com/PX4/Firmware/commit/df173d4d8a0bae59f35d38feae7e72bd06ace7f1

mhkabir commented:
@priseborough @dagar Good to merge?

LorenzMeier commented:
@mhkabir Almost!

bkueng commented:
Flight Review is updated: https://github.com/PX4/flight_review/commit/2f557c6bdeb9198241dbf27ca900605c359555cc

mhkabir commented:
@bkueng Thanks!

dagar commented:
I'm now questioning the removal of the "estimated" airspeed (=groundspeed). I think it's pretty bogus and I'm not sure if anyone is actually using it, but I also don't know if those users falling into actual airspeedless mode will be impacted. Thoughts?

priseborough commented:
The ekf2_main module can/should publish estimated airspeed using the magnitude of ground_speed -wind_speed.

dagar commented:
Sounds good. If there's no airspeed publication it can come from ekf2. My main concern was the use of plain groundspeed in place of airspeed.

mhkabir commented:
@LorenzMeier @priseborough this is good to merge now that the release is out. This also helps with our current racing focus as the bias estimates are independently published from the attitude. (good because @dagar is splitting the rate and attitude controllers soon)

priseborough commented:
OK, once this is merged I will submit a  PR so that the ekf2_main publishes an airspeed calculated using using the ground and wind speed estimates.

dagar commented:
Standby, we should probably do it all within this PR to avoid changing the
airspeed mode param at all.

On Wed, Jul 19, 2017 at 5:52 PM, Paul Riseborough <notifications@github.com>
wrote:

> OK, once this is merged I will submit a PR so that the ekf2_main publishes
> an airspeed calculated using using the ground and wind speed estimates.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PX4/Firmware/pull/7523#issuecomment-316529204>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFK6Bg0c5rsyXEqbzgOclnrYX_j5he3ks5sPnqmgaJpZM4ONAy_>
> .
>

priseborough commented:
If we rename wind_estimate.msg to airdata_estimate.msg, then in addition to the wind estimates it can be made to carry the following air data quantities:

float32 true_airspeed_m_s # true airspeed estimate (m/sec)
float32 true_airspeed_accuracy # true airspeed 1-sigma accuracy (m/sec)

@mhkabir and @dagar Let me know if you these added to the PR and I will create the required patches.

Optionally these could also be added, which are often used for control of unstable fixed wing platforms:

float32 AoA_rad # estimated angle of attack where a positive angle produces a positive Z axis wind relative velocity (rad)
float32 AoA_accuracy $ angle of attack 1-sigma accuracy (rad) 
float32 AoS_rad # estimated angle of sideslip where a positive sideslip produces a positive Y axis wind relative velocity (rad)
float32 AoS_accuracy $ angle of sideslip 1-sigma accuracy (rad)

mhkabir commented:
@priseborough, I think that makes a lot of sense. Please feel free to push directly to this branch -
 `pr-att-bias-cleanup`

@dagar What do you think?

dagar commented:
I like it, but how does this fit in with the airspeed message used by the
controllers (published when a differential pressure sensor is present)?

On Sat, Jul 22, 2017 at 10:43 PM, Kabir Mohammed <notifications@github.com>
wrote:

> @priseborough <https://github.com/priseborough>, I think that makes a lot
> of sense. Please feel free to push directly to this branch -
> pr-att-bias-cleanup
>
> @dagar <https://github.com/dagar> What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PX4/Firmware/pull/7523#issuecomment-317224493>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFK6J_DULlcn2HzXjbwF1aVePiOpCsEks5sQrNCgaJpZM4ONAy_>
> .
>

priseborough commented:
At the moment  there is no facility to use estimated data or to do useful air data fault detection and isolation. 

The preferred option in the future is that we create an air data consolidation algorithm in it's own module or ins the sensors module so that dual airspeed sensing and/or air data estimates  can be used to provide a reliable source of air data. At the moment the system is highly vulnerable to  airspeed sensor faults caused by water, foreign objects, kinked tubes and leaks. 

This is important to professional users and essential for safe for BVLOS operation.

mhkabir commented:
@priseborough since there is no immediate consumer of the airdata estimate, I see no reason to combine it with this PR. I would like this to get merged soon.

I rebased the branch. I also added some logic in mavlink to correct the `HIGHRES_IMU` readings for in-run bias.

Corrected : 
![image](https://user-images.githubusercontent.com/5557676/28507882-e2204754-7053-11e7-9abe-04194fd0a0a7.png)

Without correction (master) :
![image](https://user-images.githubusercontent.com/5557676/28508004-e4d6aad2-7054-11e7-9186-8f8f0690acac.png)

@LorenzMeier Please merge when ready.

priseborough commented:
@mhkabir I Agree.

LorenzMeier commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -2308,66 +2295,6 @@ MavlinkReceiver::handle_message_hil_state_quaternion(mavlink_message_t *msg)
 			orb_publish(ORB_ID(battery_status), _battery_pub, &hil_battery_status);
 		}
 	}
-
```
This looks like it will break HIL - I don't see the equivalents of what has been removed here being published?

LorenzMeier commented:
@bresch @sanderux Could you please both try to fly this branch? It is much needed cleanup but we need to know if VTOL is still working.

mhkabir commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -1109,29 +1059,33 @@ void Ekf2::run()
-				// Check and save the last valid calibration when we are disarmed
-				if ((vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_STANDBY)
-				    && (status.filter_fault_flags == 0)
+				// Check and update the last valid magnetometer calibration
```
@priseborough Can you please quickly sanity-check the logic here? What this now does is that it gets the _updated_ mag bias from the EKF at every step (this is for correcting the raw mag readings in `sensor_corrected`). The actual parameter _save_ is still only done when disarmed.

priseborough replied:
The EKF is already applying it's learned mag bias corrections internally, so applying them to the data continuously before it is consumed by the EKF and while the EKF runs will cause problems.

mhkabir replied:
It's not applied to the raw data (EKF behaviour is not changed in any manner). I just made the bias estimate available to other consumers without the need to land and disarm. It's just copied over to the `sensors_corrected` topic here : https://github.com/PX4/Firmware/pull/7523/files#diff-18e8f57d61a55777f8e25da991013704R965

mhkabir replied:
Note that the parameter save is still rate limited and only happens on land and disarm.

priseborough replied:
OK, TXS

mhkabir commented:
Jenkins test this please

priseborough commented:
@mhkabir I've added the airspeed to the wind_estimate as requested.

mhkabir commented:
@priseborough Thanks! Will assign @PX4TestFlights after reviewing the changes.

mhkabir commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -855,128 +846,76 @@ void Ekf2::run()
+			v_wind_comp(1) -= velNE_wind[1];
+			_vel_body_wind = R_to_body * v_wind_comp; // TODO : move this elsewhere
+
+			// use estimated velocity for airspeed estimate
```
@priseborough All of this is in the ecl EKF now, correct? Can we remove this now?

mhkabir commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -1191,38 +1165,43 @@ void Ekf2::run()
 				// Publish wind estimate
 				struct wind_estimate_s wind_estimate = {};
 				wind_estimate.timestamp = now;
 				wind_estimate.windspeed_north = status.states[22];
```
@priseborough I think it will be better if you added getters here instead of accessing the states directly like this.

mhkabir commented on src/modules/fw_att_control/fw_att_control_main.cpp:
```diff
@@ -748,12 +740,13 @@ FixedwingAttitudeControl::task_main()
 	vehicle_status_poll();
 	vehicle_land_detected_poll();
 	battery_status_poll();
+	_sub_airspeed.update();
```
@dagar Why don't we follow the same pattern here as for the topics? e.g `battery_status_poll();` etc.

dagar replied:
Yes it's ugly. Hoping to encourage someone to move the rest towards the newer APIs.

mhkabir replied:
OK, can do that after this is in.

mhkabir commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -39,6 +39,8 @@ FixedwingPositionControl *l1_control::g_control;
 static int _control_task = -1;			///< task handle for sensor task */
 
 FixedwingPositionControl::FixedwingPositionControl() :
+	_sub_airspeed(ORB_ID(airspeed), 0, 0, nullptr),
```
If I'm not wrong, this airspeed topic is generated from raw sensor data? How does this tie in with the EKF-estimated true airspeed, and why would we use the raw value? Just trying to get a better idea of how FW controllers work.

priseborough replied:
The actual airspeed is better to use if available (control loops are tuned to work with it). Estimate airspeed would only be used if an airspeed sensor was not fitted or was faulty. For airspeed estimation without a sensor, the sideslip fusion in the EKF needs to be enabled via the EKF2_AID_MASK parameter.

Investigating this capability and documenting it's performance and limitations can be done as a follow-up piece of work.

sanderux replied:
If airspeed is set to disabled it should not be used. the sensor might be bad or simply mounted for reference data. the check to apply should be done on the param, not on the valid state

mhkabir commented:
@PX4TestFlights we need testing on this, on all 3 platforms (MC,FW,VTOL).

LorenzMeier commented:
Rebased and pushed. @priseborough Could you please search for "TODO" in the ekf2 module and look at the airspeed handling? It needs consolidation.

sanderux commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -134,7 +136,7 @@ FixedwingPositionControl::parameters_update()
 	param_get(_parameter_handles.airspeed_max, &(_parameters.airspeed_max));
 	param_get(_parameter_handles.airspeed_trans, &(_parameters.airspeed_trans));
-	param_get(_parameter_handles.airspeed_mode, &(_parameters.airspeed_mode));
+	param_get(_parameter_handles.airspeed_mode, &(_parameters.airspeed_disabled));
```
mode vs disabled, could become confusing.

sanderux commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -319,37 +321,54 @@ FixedwingPositionControl::manual_control_setpoint_poll()
-	} else {
+		if (_sub_airspeed.get().indicated_airspeed_m_s > 0.0f
+		    && _sub_airspeed.get().true_airspeed_m_s > _sub_airspeed.get().indicated_airspeed_m_s) {
+			_eas2tas = max(_sub_airspeed.get().true_airspeed_m_s / _sub_airspeed.get().indicated_airspeed_m_s, 1.0f);
```
Check for disabled param

sanderux commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -319,37 +321,54 @@ FixedwingPositionControl::manual_control_setpoint_poll()
 	}
 
+	/* update TECS state */
+	_tecs.enable_airspeed(_airspeed_valid);
```
check for disabled param

mhkabir replied:
That would be a functional change which I'm happy to implement, but I will only do it if you can provide sufficient VTOL test coverage with airspeed disabled.

priseborough commented:
@LorenzMeier I've had a look through the code and the simplest way to achieve the consolidation for the first pass is to expand Sensors::diff_pres_poll so that it converts ground and wind velocity states to estimated airspeed if the differential pressure reading is not available or declared unhealthy. 

When airspeed is not available and the platform is a fixed wing type, the ekf's sideslip fusion will need to be enabled to continue wind estimation.

I also need to decide on the convention used to tell the consumer that an estimated quantity is being used. The 0 to 1 confidence field is not helpful because other than preflight checks requiring a value greater than 0.95 there is no other information on how to set or interpret this message.

priseborough commented:
Consolidating airspeed is getting too messy to incorporate with this PR. I am going to remove true airspeed calculation and handle the publication of an estimated airspeed in a separate PR.

priseborough commented:
I've cleaned up, rebased and pushed the changes here: https://github.com/PX4/Firmware/tree/pr-att-bias-cleanup2

priseborough commented:
After this PR is merged, I will create an air-data consolidator in the sensors module along with this https://github.com/PX4/Firmware/pull/7721

mhkabir commented:
Continued in #7773
